### ISSUE TITLE
Graph visualizations output can not be used on a same page

### ISSUE THREAD

**Issue by ebraminio** [igraph/python-igraph#145]
Steps to reproduce:
1. Install latest release of igraph (python-igraph-0.7.1.post6) on python2 and open a jupyter notebook (pip install jupyter), open `jupyter notebook`, create a notebook.
2. Add these line on separate cells:
```python
import igraph as ig
ig.plot(ig.Graph.Full(3), vertex_label=['a', 'b', 'c'], bbox=(100, 100))
ig.plot(ig.Graph.Full(6), vertex_label=range(1, 7), bbox=(100, 100))
```
![screen shot 2017-11-09 at 2 26 36 pm](https://user-images.githubusercontent.com/833473/32602079-07049646-c55a-11e7-8de9-0c5af06a55a4.png)

Actual:
As can be seen on the screenshot, second graph on the page is messed up. As the both are SVGs they had some interference it seems.

Expected:
Each should drawn separately and correctly.

**Comment by ebraminio** [comment]
Where can I find SVG creation related code? I have an idea about fixing the issue, lets prepend some random letters to the ids to make them unique on a page.

**Comment by ntamas** [comment]
If I remember correctly, Jupyter calls a method called `_repr_svg()` on the object returned from an expression to generate a plot. `ig.plot()` returns an object of type `Plot` in your case, so look for the `_repr_svg()` method on the `Plot` class. However, the issue that you reported seems really weird; in theory, the two `Plot` objects returned from the two calls should have no relation to each other whatsoever, so I have no idea how the first graph appears on top of the second.

Let me know if you managed to figure out something and then I'll fix it in the code. I don't use Jupyter myself so it's probably faster this way than me installing Jupyter and all its dependencies.

**Comment by ebraminio** [comment]
Hi there. Thanks for the help. I thought the drawing mechanism should be common between python and R version thats why I was searching the C package with no success this time i will go inside python sources.

The issue is they use xlink:href for labels and as svg results are embedded inside HTML, they will have common ancestor on a notebook html source. I am not sure if this should be fixed from jupyter side or igraph, or whether passing png to notebooks should be favored but adding some document constant random letters to svg ids can be an acceptable solution I guessed.

Hmm, now that I found the related https://github.com/igraph/python-igraph/commit/9cf6818ff64c2415160a32644b1a55617d7765f3 place, appending ids doesn't seem to have particular API, may be we should go with converting it to `_repr_png()`?

**Comment by ebraminio** [comment]
Hi there. Thanks for the help. I thought the drawing mechanism should be common between python and R version thats why I was searching the C package with no success this time i will go inside python sources.

The issue is they use xlink:href for labels and as svg results are embedded inside HTML, they will have common ancestor on a notebook html source. I am not sure if this should be fixed from jupyter side or igraph, or whether passing png to notebooks should be favored but adding some document constant random letters to svg ids can be an acceptable solution I guessed.

Hmm, now that I found the related https://github.com/igraph/python-igraph/commit/9cf6818ff64c2415160a32644b1a55617d7765f3 place, appending ids doesn't seem to have particular API, may be we should go with converting it to `_repr_svg()`?

**Comment by ebraminio** [comment]
Another idea could be changing jupyter notebook to use base64 encoded SVGs instead SVG-in-HTML, or, separating them into dedicated iframes.

**Comment by ebraminio** [comment]
Reported on https://github.com/jupyter/jupyter/issues/298 also

**Comment by ntamas** [comment]
> I thought the drawing mechanism should be common between python and R version
Nope; it would have been too complicated to do all the drawing stuff in C in a platform-independent way, and this way it integrates more nicely with the capabilities of the host language.

> The issue is they use xlink:href for labels and as svg results are embedded inside HTML, they will have common ancestor on a notebook html source.
Good catch!

> adding some document constant random letters to svg ids can be an acceptable solution I guessed
The catch (from igraph's side) is that we do not generate the SVG content directly - we leave it up to Cairo. I don't think there is a way to control how Cairo decides to generate the IDs in the SVG output. Using `_repr_png` instead of `_repr_svg` could be an option - it's probably only a matter of taking the current `_repr_svg` code and replacing `cairo.SVGSurface` in it with `cairo.ImageSurface`.

**Comment by ebraminio** [comment]
>  it's probably only a matter of taking the current _repr_svg code and replacing cairo.SVGSurface in it with cairo.ImageSurface.

Somehow I don't remember now why every-time I was using igraph drawing I wished it was using PNG output instead, of course the downside is blurriness, if you guess that would not be an issue I can do that, otherwise, we can wait for jupyter team response.

**Comment by ntamas** [comment]
Feel free to do it in your own copy or fork, but I'd rather wait a bit for the response of the Jupyter team to see whether they are willing to do something about it or if they know about a workaround that we could use on our side. I have subscribed to the issue that you created in the Jupyter repo.

**Comment by takluyver** [comment]
You can ask for the SVG to be isolated in an iframe by giving it metadata `{'isolated': True}`.

The `_repr_svg_` method is allowed to return a 2-tuple of `data, metadata`:

```python
return result.decode('utf-8'), {'isolated': True}
```

**Comment by ntamas** [comment]
Thanks for the info! @ebraminio, does this solve the problem?

**Comment by ebraminio** [comment]
> does this solve the problem?

Yes, it would be great if isolate could do that by base64 encode svg inside svg on jupyter side, suggested that [here](https://github.com/jupyter/jupyter/issues/298#issuecomment-344349909), but uploaded the igraph fix, thanks to @takluyver on #148
